tensornetwork/block_sparse/charge.py

```diff
@@ -264,6 +275,169 @@ def intersect(self, other, assume_unique=False, return_indices=False) -> Any:
+    """
+    obj = self.__new__(type(self))
+    tmp = np.unique(
+        self.charge_labels,
+        return_index=return_index,
+        return_inverse=return_inverse,
+        return_counts=return_counts)
+    if return_index or return_inverse or return_counts:
+      if tmp[0].ndim == 0:
+        index = np.asarray([tmp[0]])
+        unique_charges = self.unique_charges[:, index]
+      else:
+        unique_charges = self.unique_charges[:, tmp[0]]
+    else:
+      if tmp.ndim == 0:
+        tmp = np.asarray([tmp])
+      unique_charges = self.unique_charges[:, tmp]
+    obj.__init__(
+        charges=unique_charges,
+        charge_labels=np.arange(unique_charges.shape[1], dtype=np.int16),
+        charge_types=self.charge_types)
+    out = [obj]
+    if return_index or return_inverse or return_counts:
+      for n in range(1, len(tmp)):
+        out.append(tmp[n])
+
+    if len(out) == 1:
+      return out[0]
+    if len(out) == 2:
+      return out[0], out[1]
+    if len(out) == 3:
+      return out[0], out[1], out[2]
+    if len(out) == 4:
+      return out[0], out[1], out[2], out[3]
+    return None
+
+  def reduce(self,
+             target_charges: np.ndarray,
+             return_locations: bool = False,
+             strides: Optional[int] = 1) -> Any:
+    """
+    Reduce the dimension of a 
+    charge to keep only the charge values that intersect target_charges
+    Args:
+      target_charges: array of unique charges to keep.
+      return_locations: If `True`, also return the locations of 
+        target values within `BaseCharge`.
+      strides: An optional stride value.
+    Returns:
+      BaseCharge: charge of reduced dimension.
+      np.ndarray: If `return_locations = True`; the index locations 
+        of target values.
+    """
+    if isinstance(target_charges, (np.integer, int)):
+      target_charges = np.asarray([target_charges], dtype=np.int16)
+    if target_charges.ndim == 1:
+      target_charges = np.expand_dims(target_charges, 0)
+    target_charges = np.asarray(target_charges, dtype=np.int16)
+    # find intersection of index charges and target charges
+    reduced_charges, label_to_unique, _ = intersect(
+        self.unique_charges, target_charges, axis=1, return_indices=True)
+    num_unique = len(label_to_unique)
+
+    # construct the map to the reduced charges
+    map_to_reduced = np.full(self.dim, fill_value=-1, dtype=np.int16)
+    map_to_reduced[label_to_unique] = np.arange(num_unique, dtype=np.int16)
+
+    # construct the map to the reduced charges
+    reduced_ind_labels = map_to_reduced[self.charge_labels]
+    reduced_locs = reduced_ind_labels >= 0
+    new_ind_labels = reduced_ind_labels[reduced_locs].astype(np.int16)
+    obj = self.__new__(type(self))
+    obj.__init__(reduced_charges, new_ind_labels, self.charge_types)
+
+    if return_locations:
+      return obj, strides * np.flatnonzero(reduced_locs).astype(np.uint32)
```

Comment:
Why is this not int16?

Reply:
these are index positions within a large data-vector